src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/MacProxy.cs

```diff
@@ -0,0 +1,125 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Net.Http;
+using System.Net;
+using System.Collections.Generic;
+using Microsoft.Win32.SafeHandles;
+using static Interop.CoreFoundation;
+using static Interop.RunLoop;
+
+using CFRunLoopRef = System.IntPtr;
+
+namespace System.Net.Http
+{
+    internal sealed class MacProxy : IWebProxy
+    {
+        public ICredentials Credentials
+        {
+            get => null;
+            set => throw new NotSupportedException();
+        }
+
+        static Uri GetProxyUri(string scheme, CFProxy proxy)
+        {
+            var uriBuilder = new UriBuilder(
+                scheme,
+                proxy.HostName,
+                proxy.PortNumber);
+
+            // FIXME: Handle user name and password
+
+            return uriBuilder.Uri;
+        }
+
+        public Uri ExecuteProxyAutoConfiguration(SafeCreateHandle cfurl, CFProxy proxy)
+        {
+            Uri result = null;
+            CFRunLoopRef runLoop = CFRunLoopGetCurrent();
+
+            // Callback that will be called after executing the configuration script
+            CFProxyAutoConfigurationResultCallback cb = (IntPtr client, IntPtr proxyListHandle, IntPtr error) =>
+            {
+                if (proxyListHandle != IntPtr.Zero)
+                {
+                    using (var proxyList = new SafeCFArrayHandle(proxyListHandle, false))
+                    {
+                        long proxyCount = CFArrayGetCount(proxyList);
+                        for (int i = 0; i < proxyCount; i++)
+                        {
+                            IntPtr proxyValue = CFArrayGetValueAtIndex(proxyList, i);
+                            using (SafeCFDictionaryHandle proxyDict = new SafeCFDictionaryHandle(proxyValue, false))
+                            {
+                                CFProxy proxy = new CFProxy(proxyDict);
+                                if (proxy.ProxyType == CFProxy.kCFProxyTypeHTTP || proxy.ProxyType == CFProxy.kCFProxyTypeHTTPS)
+                                {
+                                    result = GetProxyUri("http", proxy);
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+                CFRunLoopStop(runLoop);
+            };
+
+            var clientContext = new CFStreamClientContext();
+            var loopSource =
+                proxy.ProxyType == CFProxy.kCFProxyTypeAutoConfigurationURL ?
+                CFNetworkExecuteProxyAutoConfigurationURL(proxy.AutoConfigurationURL, cfurl, cb, ref clientContext) :
+                CFNetworkExecuteProxyAutoConfigurationScript(proxy.AutoConfigurationJavaScript, cfurl, cb, ref clientContext);
+
+            using (var mode = CFStringCreateWithCString("System.Net.Http.MacProxy"))
```

Comment:
Nit var => whatever type it actually is

Reply:
What is the purpose of this string?  Should we use `typeof(MacProxy).FullName` instead?

Reply:
The string could be anything as long as it is sufficiently unique within the application. `typeof(MacProxy).FullName` sounds like a good option.